File: Code_on_track-daksh(ai)/Code_on_track-daksh(without ai)/Code_on_track-daksh/full_website/frontend/src/api/vendors.ts
import api from './axios';
import type { Item } from './items';

export interface Vendor {
    id: number;
    name: string;
    items_count: number;
    failed_count?: number;
    // Optional/Mapped fields
    vendor_code?: string;
    contact_name?: string;
    contact_email?: string;
    contact_phone?: string;
    address?: string;
    warranty_months?: number;
    notes?: string;
    created_at: string;
    is_active: boolean;
    components_supplied?: string[];
}

export interface VendorCreate {
    name: string;
    vendor_code: string;
    contact_name?: string;
    contact_email?: string;
    contact_phone?: string;
    address?: string;
    warranty_months?: number;
    notes?: string;
}

export interface VendorUpdate {
    name?: string;
    vendor_code?: string;
    contact_name?: string;
    contact_email?: string;
    contact_phone?: string;
    address?: string;
    warranty_months?: number;
    notes?: string;
    is_active?: boolean;
}

export interface VendorListParams {
    q?: string;
    page?: number;
    page_size?: number;
}

export const listVendors = async (params: any = {}) => {
    // Note: params can include 'id' now
    const response = await api.get<Vendor[]>('/vendors/', { params });
    return response.data;
};

export const getVendor = async (id: number) => {
    const response = await api.get<Vendor>(`/vendors/${id}`);
    return response.data;
};

export const createVendor = async (data: VendorCreate) => {
    const response = await api.post<Vendor>('/vendors', data);
    return response.data;
};

export const updateVendor = async (id: number, data: VendorUpdate) => {
    const response = await api.put<Vendor>(`/vendors/${id}`, data);
    return response.data;
};

export const deleteVendor = async (id: number) => {
    await api.delete(`/vendors/${id}`);
};

export const listVendorItems = async (id: number) => {
    const response = await api.get<Item[]>(`/vendors/${id}/items`);
    return response.data;
};

File: Code_on_track-daksh(ai)/backend/routes/vendors.py
from flask import Blueprint, jsonify, request
from sqlalchemy import text
from services.db import db

vendors_bp = Blueprint('vendors', __name__)

@vendors_bp.route('/', methods=['GET'])
def get_vendors():
    # Return vendors with items_count
    page = int(request.args.get('page', 1))
    size = int(request.args.get('page_size', 20))
    offset = (page - 1) * size

    # Get total count first
    count_sql = text("SELECT COUNT(*) FROM vendors")
    total = db.session.execute(count_sql).scalar() or 0

    sql = text("""
        SELECT v.id, v.name, v.vendor_code, v.component_supplied,
               (SELECT COUNT(*) FROM items i WHERE i.vendor_id = v.id) as items_count,
               (SELECT COUNT(*) FROM items i WHERE i.vendor_id = v.id AND i.status IN ('failed', 'rejected')) as failed_count
        FROM vendors v
        ORDER BY v.id
        LIMIT :limit OFFSET :offset
    """)
    
    try:
        rows = db.session.execute(sql, {"limit": size, "offset": offset}).fetchall()
        vendors = []
        for row in rows:
            vendors.append({
                "id": row.id,
                "name": row.name,
                "vendor_code": row.vendor_code,
                "component_supplied": row.component_supplied,
                "items_count": row.items_count,
                "failed_count": row.failed_count
            })
            
        return jsonify({
            "total": total,
            "page": page,
            "page_size": size,
            "vendors": vendors
        })
    except Exception as e:
        return jsonify({"error": str(e)}), 500

@vendors_bp.route('/<int:vendor_id>', methods=['GET'])
def get_vendor(vendor_id):
    # Fixed columns based on schema inspection:
    # warranty_months -> default_warranty_months
    # is_active -> missing in DB, assuming True or removing
    sql = text("""
        SELECT v.id, v.name, v.vendor_code, v.component_supplied, 
               v.contact_name, v.contact_email, v.contact_phone, v.address,
               v.default_warranty_months, v.notes, v.created_at,
               (SELECT COUNT(*) FROM items i WHERE i.vendor_id = v.id) as items_count,
               (SELECT COUNT(*) FROM items i WHERE i.vendor_id = v.id AND i.status IN ('failed', 'rejected')) as failed_count
        FROM vendors v
        WHERE v.id = :id
    """)
    
    try:
        row = db.session.execute(sql, {"id": vendor_id}).fetchone()
        if not row:
            return jsonify({"error": "Vendor not found"}), 404
            
        vendor = {
            "id": row.id,
            "name": row.name,
            "vendor_code": row.vendor_code,
            "component_supplied": row.component_supplied,
            "items_count": row.items_count,
            "failed_count": row.failed_count,
            "contact_name": row.contact_name,
            "contact_email": row.contact_email,
            "contact_phone": row.contact_phone,
            "address": row.address,
            "warranty_months": row.default_warranty_months, # Map back to expected key
            "notes": row.notes,
            "created_at": row.created_at.isoformat() if row.created_at else None,
            "is_active": True # Default since column missing
        }
        
        # Parse components_supplied
        comps = row.component_supplied
        if comps:
             if ',' in comps:
                 vendor["components_supplied"] = [c.strip() for c in comps.split(',')]
             else:
                 vendor["components_supplied"] = [comps]
        else:
            vendor["components_supplied"] = []

        return jsonify(vendor)
    except Exception as e:
        import traceback
        traceback.print_exc() # Log full stack trace
        return jsonify({"error": str(e)}), 500
